File: cpp/mcts/mcts.cpp
#include <memory>
#include <optional>
#include <ostream>
#include <stdexcept>
#include <iostream>
#include <functional>
#include <map>
#include <vector>

#include "../utils.cpp"

// ---------------------------------------------------------------------------------
// Games.

/*
 * Represents the state of a board game.
 */
struct GameState {
  char to_move;
  double utility;
  std::map<Index2D, char> board;
  std::vector<Index2D> moves;

};

std::ostream& operator<<(std::ostream& s, GameState state) {
  return s;
}

/*
 * Interface similar to a problem, but with utility and a terminal test
 * instead of path cost and goal test for each state.
 */

template <typename A>
struct Game {
  /*
   * List of the allowable moves at this state.
   */
  virtual std::vector<A> actions(GameState state) const = 0;

  /*
   * State that results from making a move from a state.
   */
  virtual GameState result(GameState state, A move) const = 0;

  /*
   * Returns the value of the given final state to player.
   */
  virtual double utility(GameState state, char player_mark) const = 0;

  /*
   * True if the given state is terminal for the game.
   */
  bool is_terminal(GameState state) {
    return actions(state).empty();
  }

  /*
   * Return the player whose move it is in this state.
   */
  char to_move(GameState state) {
    return state.to_move;
  }

  template <typename M>
  using PlayerFunc = std::function<M(Game<M>& game, GameState state)>;

  /*
   * Play an n-person, move-alternating game, such as TicTacToe, Chess, Go, etc.
   */
  double play_game(std::vector<PlayerFunc<A>> players) {
    auto state = initial;
    while (true) {
      for (auto const& player : players) {
        try {
          auto move = player(*this, state);
          state = result(state, move);
        } catch (std::exception& e) {}

        if (is_terminal(state)) {
          std::cout << state;
          return utility(state, to_move(initial));
        }
      }
    }
    return {};
  }

  void display(GameState state) {
    std::cout << state;
  }

  GameState initial;
};

// ---------------------------------------------------------------------------------
// Algorithm.

/*
 * Select a leaf node in the tree.
 */
template <typename A>
std::shared_ptr<MCT_Node<GameState, A>> select(std::shared_ptr<MCT_Node<GameState, A>> n) {
  if (not n->children.empty()) {
    auto comparison = [](auto n1, auto n2) { 
      auto r1 = ucb(*(n1.first));
      auto r2 = ucb(*(n2.first));
      auto result = r1 < r2;
      std::cout << r1 << " " << r2 << std::endl;
      return result; 

    };
    return select(std::max_element(n->children.begin(), n->children.end(), comparison)->first);
  }
  return n;
}
/*
 * Expand the leaf node by adding all its children states.
 */
template <typename A>
std::shared_ptr<MCT_Node<GameState, A>> expand(std::shared_ptr<MCT_Node<GameState, A>> n, Game<A>& game) {
  if (n->children.empty() and not game.is_terminal(n->state)) {
    for (const auto& action : game.actions(n->state)) {
      auto state = game.result(n->state, action);
      std::shared_ptr<MCT_Node<GameState, A>> node = std::make_shared<MCT_Node<GameState, A>>(state, n);

      n->children[node] = action;
    }

    std::cout << "\nexpand function" << std::endl;
    for (const auto& pair : n->children) {
      const auto key = pair.first;
      const A& value = pair.second;
      std::cout << "Action: " << value << ", Node Address: " << key << std::endl;
    }
  }
  return select(n);
}

/*
 * Simulate the utility of current state by random picking a step.
 */
template <typename A>
double simulate(Game<A>& game, GameState state) {
  auto player_mark = game.to_move(state);
  while (not game.is_terminal(state)) {
    auto action = random_choice(game.actions(state));
    state = game.result(state, action);
  }
  auto v = game.utility(state, player_mark);
  std::cout << "simulate utility result: " << -v << std::endl;
  return -v;
}

/*
 * Passes the utility back to all parent nodes.
 */
template <typename A>
void backprop(std::shared_ptr<MCT_Node<GameState, A>> n, double utility) {
  if (utility > 0) {
    n->U += utility;
  }
  n->N++;
  if (n->parent != nullptr) {
    backprop(n->parent, -utility);
  }
}

template <typename A>
A monte_carlo_tree_search(GameState state, Game<A>& game, int N=1000) {
  std::shared_ptr<MCT_Node<GameState, A>> root = std::make_shared<MCT_Node<GameState, A>>(state);
  for (size_t i{}; i < 1000; i++) {
    auto leaf = select(root);

    std:: cout << "\nselect leaf" << std::endl;
    for(auto index2d : leaf->state.board) {
      std::cout << index2d.first << std::endl;
    }

    auto child = expand(leaf, game);

    for(auto index2d : child->state.board) {
      std::cout << index2d.first << std::endl;
    }

    // auto keys = root->children.begin(); return keys->second;
    //
    auto result = simulate(game, child->state);
    backprop(child, result);
  }

  std::cout << "\nroot childrens" << std::endl;
  for (const auto& pair : root->children) {
    const auto key = pair.first;
    const A& value = pair.second;
    std::cout << "N: " << key->N << ", U: " << key->U << ", Action: " << value << ", Node Address: " << key << std::endl;
  }

  auto comparison = [](auto p1, auto p2){ return p1.first->N < p2.first->N; };
  return std::max_element(root->children.begin(), root->children.end(), comparison)->second;
}

// ---------------------------------------------------------------------------------
// Players.

template <typename A>
A random_player(Game<A>& game, GameState state) {
  return random_choice(game.actions(state));
}

template <typename A>
A mcts_player(Game<A>& game, GameState state) {
  return monte_carlo_tree_search(state, game);
}

// ---------------------------------------------------------------------------------
// Game implementations.

struct TicTacToe : Game<Index2D> {
  TicTacToe(int h = 3, int v = 3, int k = 3)
  : Game<Index2D>{},
    h{ h },
    v{ v },
    k{ k } {
    std::vector<Index2D> moves;
    for (size_t x = 1; x < h + 1; x++) {
      for (size_t y = 1; y < v + 1; y++) {
        moves.emplace_back(x, y);
      }
    }
    initial = { 'X', 0, {}, moves };
  }

  std::vector<Index2D> actions(GameState state) const override  {
    return state.moves;
  }

  GameState result(GameState state, Index2D move) const override {
    if (not is_in(state.moves, move)) {
      return state;
    }
    auto board = state.board;
    board[move] =  state.to_move;
    auto moves = state.moves;
    remove(moves, move);
    return GameState{
      state.to_move == 'X' ? 'O' : 'X',
      compute_utility(board, move, state.to_move),
      board,
      moves
    };
  }

  double utility(GameState state, char player_mark) const override {
    return player_mark == 'X' ? state.utility : -state.utility;
  }

  bool is_terminal(GameState state) {
    return state.utility != 0 or state.moves.size() == 0;
  }

  void display(GameState state) {
    auto board = state.board;
    for (int x = 1; x < h + 1; x++) {
      for (int y = 1; y < h + 1; y++) {
        auto it = board.find({ x, y });
        auto el = it == board.end() ? '.' : it->second;
        std::cout << el << " ";
      }
      std::cout << "\n";
    }
  }

  double compute_utility(std::map<Index2D, char> board, Index2D move, char player_mark) const {
    if (k_in_row(board, move, player_mark, {0, 1}) or
        k_in_row(board, move, player_mark, {1, 0}) or
        k_in_row(board, move, player_mark, {1, -1}) or
        k_in_row(board, move, player_mark, {1, 1})) {
      return player_mark == 'X' ? 1 : -1;
    }
    return 0;
  }

  bool k_in_row(std::map<Index2D, char> board, Index2D move, char player_mark, Index2D delta_x_y) const {
    auto [delta_x, delta_y] = delta_x_y;
    auto [x, y] = move;
    // number of moves in row.
    int n = 0;

    while (board.find({ x, y })->second == player_mark) {
      n += 1;
      x += delta_x;
      y += delta_y;
    }
    x = move.first;
    y = move.second;

    while (board.find({ x, y })->second == player_mark) {
      n += 1;
      x += delta_x;
      y += delta_y;
    }
    n--; // because we counted move itself twice.
    return n >= k;
  }

  int h;
  int v;
  int k;
};

File: js/mcts/mcts.js
"use strict";

// ----------------------------------------------------------------------------
// Utils.

class MCTNode {
  static lastAssignedId = 0;

  constructor(state = null, parent = null, U = 0, N = 0) {
    this.parent = parent;
    this.state = state;
    this.U = U;
    this.N = N;
    this.children = new Map();
    this.actions = null;
    MCTNode.lastAssignedId += 1;
    this.id = MCTNode.lastAssignedId;
  }
}

class GameState {
  constructor(to_move, utility, board = new Map(), moves = []) {
    this.to_move = to_move;
    this.utility = utility;
    // This will be a Map from stringified 2D indexes to strings representing the player.
    this.board = board;
    this.moves = moves;
  }
}

function ucb(n, C = 1.4) {
  return n.N == 0 ? Infinity : n.U / n.N + C * Math.sqrt(Math.log(n.parent.N) / n.N)
}

function isEmpty(it) {
  for (const i of it) return false;
  return true;
}

function maxBy(it, keyFunc) {
  let arr = [...it];
  if (arr.length === 0) {
    return undefined;
  }
  return arr.reduce((max, current) => (keyFunc(current) > keyFunc(max) ? current : max), arr[0]);
}

// Checks if an array is in an iterable. This assumes the iterable has arrays as
// its elements.
function isIn(arr, iterable) {
  for (const x of iterable) {
    if (x.every((val, idx) => val === arr[idx])) {
      return true;
    }
  }
  return false;
}

function randomChoice(arr) {
  if (arr.length === 0) {
    return undefined;
  }

  const randomIndex = Math.floor(Math.random() * arr.length);
  return arr[randomIndex];
}

// ----------------------------------------------------------------------------
// Monte Carlo Tree Search.

function monteCarloTreeSearch(state, game, N = 1000) {
  const select = (n) => {
    if (!isEmpty(n.children))
      return select(maxBy(n.children.keys(), ucb));
    return n;
  };

  const expand = (n) => {
    if (isEmpty(n.children) && !game.terminal_test(n.state)) {
      n.children = new Map();
      game.actions(n.state).forEach((action) => {
        const newNode = new MCTNode(game.result(n.state, action), n);
        n.children.set(newNode, action);
      })
    }
    return select(n);
  };

  const simulate = (game, state) => {
    const player = game.to_move(state);
    while (!game.terminal_test(state)) {
      const action = randomChoice(game.actions(state));
      state = game.result(state, action);
    }
    const v = game.utility(state, player);
    return -v;
  };

  const backprop = (n, utility) => {
    if (utility > 0)
      n.U += utility;
    n.N += 1;
    if (n.parent != null)
      // Negates because the tree is alternated.
      backprop(n.parent, -utility);
  };

  let root = new MCTNode(state);

  for (let i = 0; i < N; i++) {
    const leaf = select(root);
    const child = expand(leaf);
    const result = simulate(game, child.state);
    backprop(child, result);
  }

  const max_state = maxBy(root.children.keys(), (p) => p.N);
  return root.children.get(max_state);
}

// ----------------------------------------------------------------------------
// Games.

/*
 * Interface for an abstract game.
 */
class Game {
  actions(state) {
    return undefined;
  }

  result(state, move) {
    return undefined;
  }

  utility(state, player) {
    return undefined;
  }

  terminal_test(state) {
    return this.actions(state).length == 0;
  }

  to_move(state) {
    return state.to_move;
  }
}

class TicTacToe extends Game {
  constructor(h = 3, v = 3, k = 3) {
    super();
    this.h = h;
    this.v = v;
    this.k = k;
    
    const moves = [];
    for (let x = 1; x <= h; x++) {
      for (let y = 1; y <= v; y++) {
        moves.push([x, y]);
      }
    }

    this.initial = new GameState('X', 0, new Map(), moves);
  }

  actions(state) {
    return state.moves;
  }

  result(state, move) {
    if (!state.moves.some(el => el[0] == move[0] && el[1] == move[1])) // Intankavel o JavaScript
      return state;
    const board = new Map(state.board);
    board.set(move.toString(), state.to_move);
    let moves = Array.from(state.moves);
    moves = moves.filter(x => !x.every((val, idx) => val == move[idx]));
    return new GameState(
      state.to_move === 'X' ? 'O' : 'X',
      this.compute_utility(board, move, state.to_move),
      board,
      moves
    );
  }

  utility(state, player) {
    return player === 'X' ? state.utility : -state.utility;
  }

  terminal_test(state) {
    return state.utility != 0  || state.moves.length === 0;
  }

  compute_utility(board, move, player) {
    if (this.k_in_row(board, move, player, [0, 1]) ||
        this.k_in_row(board, move, player, [1, 0]) ||
        this.k_in_row(board, move, player, [1,-1]) ||
        this.k_in_row(board, move, player, [1, 1]))
      return player === 'X' ? 1 : -1;
    else
      return 0
  }

  k_in_row(board, move, player, delta_x_y) {
    const [delta_x, delta_y] = delta_x_y;
    let [x, y] = move;
    let n = 0;
    while (board.get([x, y].toString()) == player) {
      n += 1;
      [x, y] = [x + delta_x, y + delta_y];
    }
    [x, y] = move;
    while (board.get([x, y].toString()) == player) {
      n += 1;
      [x, y] = [x - delta_x, y - delta_y];
    }
    n -= 1
    return n >= this.k;
  }
}

// Tests

const ttt = new TicTacToe();

function gen_state(to_move = 'X', x_positions = [], o_positions = [], h = 3, v = 3) {
  let moves = []
  for (let x = 1; x <= h; x++) {
    for (let y = 1; y <= v; y++) {
      moves.push([x, y]);
    }
  }
  const xSet = new Set(x_positions);
  const oSet = new Set(o_positions);
  moves = moves.filter(pair => !isIn(pair, xSet) && !isIn(pair, oSet));
  const board = new Map();
  for (const pos of x_positions)
    board.set(pos.toString(), 'X');
  for (const pos of o_positions)
    board.set(pos.toString(), 'O');
  return new GameState(to_move, 0, board, moves);
}

function test_monte_carlo_tree_search() {
  // Plays on middle if it guarantees a win.
  let state = gen_state(
    'X',
    [[1, 1], [3, 3]], // X positions.
    [[1, 2], [3, 2]]  // O positions.
  );
  let result = monteCarloTreeSearch(state, ttt).toString();
  console.log(result == '2,2');

  // Plays on either position to win.
  state = gen_state(
    'X',
    [[1,1], [3, 3], [3, 1]],
    [[1,2], [3, 2]]
  );
  result = monteCarloTreeSearch(state, ttt).toString();
  console.log(['2,2', '2,1'].includes(result));

  // Plays to avoid enemy's win.
  state = gen_state(
    'X',
    [[1, 1]],
    [[1, 2], [3, 2]]
  );
  result = monteCarloTreeSearch(state, ttt).toString();
  console.log(result == '2,2');
}

test_monte_carlo_tree_search();
